### PR TITLE
[1st version] Add Tuple idl

### PR DESCRIPTION
**Pull request by Mecoli1219**
## Tracking issue
https://github.com/flyteorg/flyte/issues/5908
https://github.com/flyteorg/flyte/issues/1337
https://github.com/flyteorg/flyte/issues/3158
https://github.com/flyteorg/flyte/issues/4358

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
RFC: https://github.com/flyteorg/flyte/pull/5699

### CODE REVIEW

File: flyteidl/clients/go/coreutils/literals.go [hunk 1]
```diff
@@ -637,6 +667,47 @@ func MakeLiteralForType(t *core.LiteralType, v interface{}) (*core.Literal, erro
...
+		//
+		// [Example usage]
+		// inputs:
+		//     tuple_data:
+		//         key1: "foo"
+		//         key2: 123
+
+		vMap, ok := v.(map[string]interface{})
+		if !ok {
+			return nil, errors.Errorf("Expected a map[string]interface{} for tuple type, got [%v]", v)
```
Comment (Future-Outlier) [inline flyteidl/clients/go/coreutils/literals.go:681]: Can you explain more?

Comment (Mecoli1219) [reply flyteidl/clients/go/coreutils/literals.go:681]: In this part, the LiteralType is already given, so the name of the tuple and the order of each field should be already defined. The only thing that is missing is the value of each field.
Therefore, we only need to provide the value of each tuple field (via the key-value pair of each field) in `flytectl` for the registered task or workflow for further execution needs.

File: flyteidl/protos/flyteidl/core/literals.proto [hunk 2]
```diff
@@ -140,6 +143,18 @@ message LiteralMap {
...
+// A collection of fields for tuple literal.
+message LiteralTupleMap {
+    // The name of the NamedTuple. If it is original tuple, it would be empty string.
+    string tuple_name = 1;
+
+    // The order of each fields stored in the tuple.
+    repeated string order = 2;
+    
+    // A map of literals.
+    map<string, Literal> literals = 3;
```
Comment (pingsutw) [inline flyteidl/protos/flyteidl/core/literals.proto:155]: Can we use `Literal.collection`?

File: flyteidl/protos/flyteidl/core/types.proto [hunk 3]
```diff
@@ -108,6 +108,14 @@ message UnionType {
     repeated LiteralType variants = 1;
 }
 
+// Defines a named tuple type
+// TODO: Add more docs
+message TupleType {
+    string tuple_name = 1;
+    repeated string order = 2;
+    map<string, LiteralType> fields = 3;
```
Comment (pingsutw) [inline flyteidl/protos/flyteidl/core/types.proto:116]: Can we use repeated here, like

```proto
message TupleEntry {
  string name = 1;
  LiteralType literal_type = 2;
}
message TupleType {
    string tuple_name = 1;
    repeated TupleEntry = 2;
}
```

File: flyteidl/protos/flyteidl/core/literals.proto (hunk 2 above)
Comment (Mecoli1219) [reply flyteidl/protos/flyteidl/core/literals.proto:155]: `Literal.collection` is not enough to store the name of the fields for `NamedTuple`.

File: flyteidl/protos/flyteidl/core/types.proto (hunk 3 above)
Comment (Mecoli1219) [reply flyteidl/protos/flyteidl/core/types.proto:116]: > Other Considerations
An alternative approach to storing the literal values for NamedTuple and Tuple would be combining the order and the map into a list of messages, as shown below:

// !!! This is not the proposed implementation !!!
message FakeLiteralTupleField {
    string name = 1;
    Literal value = 2;
}

message FakeLiteralTupleMap {
    string tuple_name = 1;
    repeated FakeLiteralTupleField fields = 2;
}
While this structure may seem more straightforward, it has a drawback: looking up a field’s value by name requires iterating through the list, which results in O(n) complexity. In contrast, using a map allows for more efficient O(1) lookup times when retrieving the value of a field by its name. Although in practice the tuple size is usually small, the map-based approach is more scalable and efficient.

**Pull request closed without merging**